uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/absence-period-model_test.js

```diff
@@ -0,0 +1,114 @@
+define([
+  'leave-absences/shared/models/absence-period-model',
+  'mocks/apis/absence-period-api-mock',
+  'common/mocks/services/hr-settings-mock',
+], function () {
+  'use strict'
+
+  describe('AbsencePeriod', function () {
+    var $provide, AbsencePeriod, AbsencePeriodAPI, $rootScope;
+
+    beforeEach(module('leave-absences.models', 'leave-absences.mocks', 'common.mocks',
+      function (_$provide_) {
+        $provide = _$provide_;
+      }));
+
+    beforeEach(inject(function (_AbsencePeriodAPIMock_, _HR_settingsMock_) {
+      $provide.value('AbsencePeriodAPI', _AbsencePeriodAPIMock_);
+      $provide.value('HR_settings', _HR_settingsMock_);
+    }));
+
+    beforeEach(inject(function (_AbsencePeriod_, _AbsencePeriodAPI_, _$rootScope_) {
+      AbsencePeriod = _AbsencePeriod_;
+      AbsencePeriodAPI = _AbsencePeriodAPI_;
+      $rootScope = _$rootScope_;
+
+      spyOn(AbsencePeriodAPI, 'all').and.callThrough();
+    }));
+
+    it('has expected interface', function () {
+      expect(Object.keys(AbsencePeriod)).toEqual(['all', 'current']);
+    });
+
+    describe('all()', function () {
+      var promise;
+
+      beforeEach(function () {
+        promise = AbsencePeriod.all();
+      });
+
+      afterEach(function () {
+        //to excute the promise force an digest
+        $rootScope.$apply();
+      });
+
+      it('calls equivalent API method', function () {
+        promise.then(function (response) {
+          expect(AbsencePeriodAPI.all).toHaveBeenCalled();
+        });
+      });
+
+      it('returns model instances', function () {
+        promise.then(function (response) {
+          expect(response.every(function (modelInstance) {
+            return 'init' in modelInstance;
+          })).toBe(true);
+        });
+      });
+    });
+
+    describe('current()', function () {
+      var promise;
+
+      beforeEach(function () {
+        promise = AbsencePeriod.current();
+      });
+
+      afterEach(function () {
+        //to excute the promise force an digest
+        $rootScope.$apply();
+      });
+
+      it('calls equivalent API method', function () {
+        promise.then(function (response) {
+          expect(AbsencePeriodAPI.all).toHaveBeenCalled();
+        });
+      });
+
+      it('returns model instance', function () {
+        promise.then(function (response) {
+          expect(response.init).toBeDefined();
+        });
+      });
+
+      describe('with past absence periods', function () {
```

Comment:
This entire test structure seems extremely overcomplicated.

`.current()`'s functionality is to return the absence period that the *current* date is in. So why now simply change the current date in a couple of different tests and then test the result? In jasmine you can change the current time by doing:
```js
jasmine.clock().mockDate(/* A Date object */);
```
it will make the whole test easier: just control what date it is in the test, and based on that define your expectations (given that you also know what the mocked data contains in terms of absence periods dates)